ext/liquid_c/block.c

```diff
@@ -30,25 +33,68 @@ typedef struct parse_context {
     VALUE ruby_obj;
 } parse_context_t;
 
+typedef struct internal_block_body {
+    bool compiled;
+    VALUE obj;
+
+    union {
+        struct {
+            VALUE document_body_obj;
+            c_buffer_t *buffer;
+            size_t offset;
+        } compiled;
+        struct {
+            VALUE source;
+            bool blank;
+            bool root;
+            int render_score;
+            vm_assembler_t code;
+        } intermediate;
+    } as;
+} internal_block_body_t;
+
+static block_body_t *block_body_ptr(const internal_block_body_t *body)
+{
+    return ((block_body_t *)(body->as.compiled.buffer->data + body->as.compiled.offset));
+}
+
 static void block_body_mark(void *ptr)
 {
-    block_body_t *body = ptr;
-    rb_gc_mark(body->source);
-    vm_assembler_gc_mark(&body->code);
+    internal_block_body_t *body = ptr;
+    if (body->compiled) {
+        rb_gc_mark(body->as.compiled.document_body_obj);
+        block_body_t *block_body = block_body_ptr(body);
+        VALUE *start = block_body_constants_ptr(block_body);
+        assert(block_body->constants_bytes % sizeof(VALUE *) == 0);
+        size_t len = block_body->constants_bytes / sizeof(VALUE *);
+        for (size_t i = 0; i < len; i++) {
+            rb_gc_mark(start[i]);
+        }
+    } else {
+        rb_gc_mark(body->as.intermediate.source);
+        vm_assembler_gc_mark(&body->as.intermediate.code);
+    }
 }
 
 static void block_body_free(void *ptr)
 {
-    block_body_t *body = ptr;
-    vm_assembler_free(&body->code);
+    internal_block_body_t *body = ptr;
+    if (!body->compiled) {
+        vm_assembler_free(&body->as.intermediate.code);
+    }
     xfree(body);
 }
 
 static size_t block_body_memsize(const void *ptr)
 {
-    const block_body_t *body = ptr;
+    const internal_block_body_t *body = ptr;
     if (!ptr) return 0;
-    return sizeof(block_body_t) + vm_assembler_alloc_memsize(&body->code);
+    if (body->compiled) {
+        block_body_t *block_body = block_body_ptr(body);
+        return sizeof(internal_block_body_t) + block_body->instructions_bytes + block_body->constants_bytes;
```

Comment:
Isn't the instruction and constant bytes a part of the document body at this point?  So I think you should count that memory in the document body object.